embedded cover extraction from ogg file, using ffmpeg

jixunmoe opened the pull request:
OGG container does not use jpeg nor png for its cover - it's a video stream encoded in `theora` (check attachment for a sample ogg with cover):

```
Input #0, ogg, from 'test_with_cover.ogg':
  Duration: 00:00:03.66, start: 0.000000, bitrate: 22 kb/s
  Stream #0:0: Video: theora, yuv444p, 214x152 [SAR 1:1 DAR 107:76], 25 tbr, 25 tbn, 25 tbc
    Metadata:
      encoder         : Lavc59.37.100 libtheora
  Stream #0:1: Audio: vorbis, 44100 Hz, stereo, fltp, 112 kb/s
    Metadata:
      encoder         : Lavf59.27.100
```

ffmpeg metadata provider asks ffmpeg to "copy" the stream, which ends up with a file that could not be identified / parsed in the browser, a "broken" image.

I could not find any integration test that actually invokes ffmpeg, no test was added.

[test_with_cover.ogg.gz](https://github.com/navidrome/navidrome/files/10798539/test_with_cover.ogg.gz)

jixunmoe committed:
fix: convert cover to jpeg if video stream was not jpeg nor png

dweymouth commented on core/ffmpeg/ffmpeg.go:
```diff
@@ -27,7 +27,9 @@ func New() FFmpeg {
-	extractImageCmd = "ffmpeg -i %s -an -vcodec copy -f image2pipe -"
+	// OGG cover uses "Theora", which is not supported by the browser's image tag.
+	// However, we can tell ffmpeg to perform the image conversion for us.
+	extractImageCmd = "ffmpeg -i %s -an -c:v mjpeg -f image2pipe -"
```
I don't know ffmpeg commands super well but does this convert all covers to jpeg, or only theora? I assume we don't want to convert png to jpg to avoid quality loss since png is supported everywhere

jixunmoe replied:
It will always converted to jpg. To conditionally copy/convert, we need to know the video stream content beforehand and decide which codec to use (`copy` vs `mjpeg`).

dweymouth replied:
Perhaps it should be rewritten then to try the mjpeg extraction only if the copy extraction fails? I think it's a regression of existing behavior if navidrome previously extracted embedded PNG art without converting it and now it won't.

jixunmoe replied:
Make sense. The copy won't fail in ogg's case though - it extracts and serves the "raw" theora which was not identifiable by Chrome.

We could cache the video stream type during probing, or run the probing again on the fly when we need to extract the image.

I can give this a try tomorrow.

jixunmoe replied:
Changed to detect image streams.

In the case of jpg/png/webp, it will extract as it is. Otherwise, it will be converted to jpg.

jixunmoe committed:
feat: probe and convert cover only when necessary

- In the case of jpg/png/webp cover, the video stream will be extracted as it is.
- In the case of non-image (e.g. theora, x264), it will be converted to mjpeg (jpg).

jixunmoe committed:
chore: format test code, oops

jixunmoe commented:
Pipeline build: https://github.com/jixunmoe/navidrome/actions/runs/4277153559

metalheim commented on core/ffmpeg/ffmpeg.go:
```diff
@@ -27,12 +29,19 @@ func New() FFmpeg {
+	//   Stream #0:0: Video: png, rgb24(pc),
+	videoStreamTypeRegex   = regexp.MustCompile(`Stream #\d+:\d*(?:\(\w+\))?: Video: ([^\s,]+)`)
+	supportedRawStreamType = []string{"webp", "mjpeg", "png"}
+)
```
gif should be supported on all clients (Web, Android, ios) natively. Maybe also avif, not sure. 

Consider putting this hardcoded list into an environment variable, so its changeable later on?

jixunmoe replied:
Not sure about having this as a config - it feels like unneeded level of complexity to preserve the raw image stream that may or may not be supported in users' device.

To be honest I'm not even sure if `webp` was a good choice considering external media player support...

dweymouth replied:
Not all clients support Gif, at least not without knowing ahead of time that the image is a Gif (because they may have to use a different codepath to support it), and the Subsonic API unfortunately doesn't have a format attribute on the image assets.

deluan commented on core/ffmpeg/ffmpeg.go:
```diff
@@ -45,10 +56,25 @@ func (e *ffmpeg) ExtractImage(ctx context.Context, path string) (io.ReadCloser,
 	}
-	args := createFFmpegCommand(extractImageCmd, path, 0)
+
+	imgFormat, err := e.GetVideoStreamType(ctx, path)
```
I'm not fan of increasing memory and cpu consumption to handle an edge case like this, calling ffmpeg twice for each image. Can't we cameup with a "smart" ffmpeg command to do that in one call?

jixunmoe replied:
The first call will not involve any conversion, as it just asks `ffmpeg` to read & parse the metadata of the media file.

Alternatively, we can ask `ffmpeg` to extract the raw stream like before, check for known image magic header, then ask `ffmpeg` to convert the stream if we don't recognize it.

jixunmoe replied:
Re-implemented the logic (integration test needed), asking ffmpeg to convert the file when we don't know what the buffer is.
Happy path - everything is fast. Sad path - ffmpeg is reading the audio file twice.

deluan reviewed (commented):
Thanks for this! Please see my comment below.

jixunmoe committed:
refactor: revert changes and get ready to implement the new detection logic

jixunmoe committed:
feat: implement to logic to extract jpg when initial dump was not a recognized image.

jixunmoe committed:
fix: lint issue

jixunmoe committed:
Merge remote-tracking branch 'upstream/master'

jixunmoe committed:
test: fix js test when ran against node 18

jixunmoe committed:
Merge remote-tracking branch 'upstream/master'
